### PR TITLE
Add memory parser

### PR DESCRIPTION
**Pull request by cholcombe973**
I added a parser for the memory device. It mostly follows your examples in the other modules. I took a slightly different approach with parsing the individual fields though. I verified the parsing against some smbios 3.0 servers I have at work. Seems to work ok.

### CODE REVIEW

File: src/memory.rs [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+            )));
+            memory.firmware_version = Some(find_string(&structure, &mut mem_pointer)?);
+            memory.module_manufacturer = Some(get_word(&mut mem_pointer, &structure.data));
+            memory.module_product_id = Some(get_word(&mut mem_pointer, &structure.data));
+            memory.memory_subsystem_controller_manufacturer_id =
+                Some(get_word(&mut mem_pointer, &structure.data));
+            memory.memory_subsystem_controller_product_id =
+                Some(get_word(&mut mem_pointer, &structure.data));
+            memory.non_volatile_size =
+                get_optional_qword(&mut mem_pointer, &structure.data, 18446744073709551615);
```
[the remark this answers is not in the document]
Comment (cholcombe973) [reply src/memory.rs:437]: Yeah no problem

[the remark this answers is not in the document]
Comment (cholcombe973) [reply src/memory.rs]: Would you like the `From` impl's to use hex notation as well?

File: src/memory.rs [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    let word = get_word(pointer, data);
+    if word == none_val {
+        None
+    } else {
+        Some(word)
+    }
+}
+
+fn get_word(pointer: &mut usize, data: &[u8]) -> u16 {
+    let word = u16::from_le_bytes(data[*pointer..(*pointer + 2)].try_into().unwrap());
```
Comment (cholcombe973) [inline src/memory.rs:483]: I'm not familiar with how the fail crate works so I used unwrap here but I'd like to remove that if possible

File: src/memory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcreekmore) [reply src/memory.rs]: Nah, the small numbers in the `From` impls are fine. It was just those giant constants that weren't clearly all Fs.

File: src/memory.rs (hunk 2 above)
Comment (cholcombe973) [reply src/memory.rs:483]: I'm going to remove these unwrap's so hang tight. I'll push another commit in a few mins

Comment (jcreekmore) [reply src/memory.rs:483]: Will do.

**Pull request merged**